### DISCUSSION TITLE
Pasting issues total crash

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#58] [General]
I was pasting in a f77 source, and a rather small paste buffer of 20 or so lines (I'll have to check the screenshots on the laptop) of the first program and comment block up to the name of pi4.f from pitest (NASA), which I've used as a benchmark for decades.  This was in iaf in putty using the TEXT command.  Pasting that buffer kills dtcyber outright.  I'll add screenshots to this post after I get on my laptop.  Should be repeatable, was for me.

**Comment by kej715** [comment]
Not able to reproduce so far, even when attempting to paste a text file of
almost 200 lines. The large file causes IAF to report typeahead overflow,
and it drops out of TEXT mode, but DtCyber does not crash. Please supply a
copy of the f77 source you are attempting to paste, along with information
about how you have configured putty to communicate with DtCyber (e.g., any
non-default Telnet settings?).

FYI ... NOS 2.8.7 supports Kermit, XMODEM, and FTP. These might be better
options for uploading source files than attempting to use copy/paste with
TEXT mode, especially given that TEXT mode doesn't provide any flow
control, and my tests show that its typeahead capability can be overrun..

thanks,
Kevin

**Comment by kej715** [comment]
I've been able to reproduce the problem. Putty is running DtCyber out of input buffers, and buffer exhaustion is handled as a fatal error. The number of buffers pre-allocated by DtCyber is 1000, so this probably indicates that putty is sending one character per Telnet packet. Possibly, DtCyber can be enhanced to coalesce short packets to avoid the problem. However, as previously suggested, using a file transfer application such as FTP is probably a better and more reliable way to upload bulk text than trying to paste into IAF text mode. Even if DtCyber was enhanced to avoid the buffer exhaustion problem, IAF TEXT mode would probably abort due to typeahead overrun.

**Comment by KC5VDJ** [comment]
I agree on the file transfer thing..  been 34 years since I laid hands on a
real cyber, and since we were BE then SE (was USN PACMISTESTCEN the ONLY
NOS-SE site?  Mr. Tracy (CDC Eng) did it local,  so cyber ftp is kinda new
to me.  (SE == "Secure Environment").

So, it is dtcyber and not NOS 2?  I've been worried about buffer overrun
issues, mostly though in the NOS code or the network processor code,
especially after seeing how ram constrained they were despite the 24 bit
addressing of the 68k.  Is the tcp/ip safe from basic Morris worm shit
(happened while I had that job)?  I'd like to do something internet facing,
but I have concerns...  Three protocols I'd like to implement for nos just
for shits and giggles too, so I will be digging in...

As for the single char per packet thing, there was a discussion about that,
iirc, in one of the meeting notes in the IENs, about the time of the cut
from the IMPs.  I forget which though, but it was after WarGames, which is
also mentioned as a concern in the same or adjacent meeting notes IENs.
I'll have to look that up.  Maybe they were thinking something similar, but
to do it otherwise involves compromises, and timeouts, and timeouts could
effect other stuff, waiting n ticks for a packet to maybe fill before
sending any of it...response time issues using both methods, but for
conversational modes one char at a time for both in, echo, and out seems
natural.  Filling a packet first is best for block modes, I would think.

One of these days I'll dig into putty.  It needs some terminal emulations
though, and I have hundreds of terminal manuals.

Hell, let's do a vote.  Most hideously poorly laid out terminal keyboard.
I vote Hewlett Packard.  Tandem is a distant second.

**Comment by rly12001** [comment]
Your message dated: Sat, 15 Jul 2023 10:59:20 -0700
--------

Hi:

	Not to intrude too much.... but you may be able to send multiple
	bytes per packet using "local line editing" to force on in putty.

	The following stackoverflow thread tells how to do it with serial
	comm lines, but maybe it would work over tcp as well.

	https://stackoverflow.com/questions/4999280/how-to-send-characters-in-putty-serial-communication-only-when-pressing-enter

	-ron

>
> I agree on the file transfer thing..  been 34 years since I laid hands on=
>  a
> real cyber, and since we were BE then SE (was USN PACMISTESTCEN the ONLY
> NOS-SE site?  Mr. Tracy (CDC Eng) did it local,  so cyber ftp is kinda ne=
> w
> to me.  (SE =3D=3D "Secure Environment").
>
> So, it is dtcyber and not NOS 2?  I've been worried about buffer overrun
> issues, mostly though in the NOS code or the network processor code,
> especially after seeing how ram constrained they were despite the 24 bit
> addressing of the 68k.  Is the tcp/ip safe from basic Morris worm shit
> (happened while I had that job)?  I'd like to do something internet facin=
> g,
> but I have concerns...  Three protocols I'd like to implement for nos jus=
> t
> for shits and giggles too, so I will be digging in...
>
> As for the single char per packet thing, there was a discussion about tha=
> t,
> iirc, in one of the meeting notes in the IENs, about the time of the cut
> from the IMPs.  I forget which though, but it was after WarGames, which i=
> s
> also mentioned as a concern in the same or adjacent meeting notes IENs.
> I'll have to look that up.  Maybe they were thinking something similar, b=
> ut
> to do it otherwise involves compromises, and timeouts, and timeouts could=
>
> effect other stuff, waiting n ticks for a packet to maybe fill before
> sending any of it...response time issues using both methods, but for
> conversational modes one char at a time for both in, echo, and out seems
> natural.  Filling a packet first is best for block modes, I would think.
>
> One of these days I'll dig into putty.  It needs some terminal emulations=
>
> though, and I have hundreds of terminal manuals.
>
> Hell, let's do a vote.  Most hideously poorly laid out terminal keyboard.=
>
> I vote Hewlett Packard.  Tandem is a distant second.
>
> On Sun, Jul 2, 2023, 12:20 Kevin Jordan ***@***.***> wrote:
>
> > I've been able to reproduce the problem. Putty is running DtCyber out o=
> f
> > input buffers, and buffer exhaustion is handled as a fatal error. The
> > number of buffers pre-allocated by DtCyber is 1000, so this probably
> > indicates that putty is sending one character per Telnet packet. Possib=
> ly,
> > DtCyber can be enhanced to coalesce short packets to avoid the problem.=
>
> > However, as previously suggested, using a file transfer application suc=
> h as
> > FTP is probably a better and more reliable way to upload bulk text than=
>
> > trying to paste into IAF text mode. Even if DtCyber was enhanced to avo=
> id
> > the buffer exhaustion problem, IAF TEXT mode would probably abort due t=
> o
> > typeahead overrun.
> >
> > =E2=80=94
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kej715/DtCyber/discussions/58#discussioncomment-633=
> 7991>,
> > or unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/BAPCTQ776WMUJD3ZUM7B=
> EB3XOG3WNANCNFSM6AAAAAAZ26VCHM>
> > .
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
>
>
> -- =
>
> Reply to this email directly or view it on GitHub:
> https://github.com/kej715/DtCyber/discussions/58#discussioncomment-645721=
> 1
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***=
> om>=
>
> ----==_mimepart_64b2de78a6ae9_1364c7d0704227ec
> Content-Type: text/html;
>  charset=UTF-8
> Content-Transfer-Encoding: quoted-printable
>
> <p></p>
> I agree on the file transfer thing..  been 34 years since I laid hands on=
>  a
> <br>
> real cyber, and since we were BE then SE (was USN PACMISTESTCEN the ONLY
> <br>
> NOS-SE site?  Mr. Tracy (CDC Eng) did it local,  so cyber ftp is kinda ne=
> w
> <br>
> to me.  (SE =3D=3D &quot;Secure Environment&quot;).
> <br>
>
> <br>
> So, it is dtcyber and not NOS 2?  I&#39;ve been worried about buffer over=
> run
> <br>
> issues, mostly though in the NOS code or the network processor code,
> <br>
> especially after seeing how ram constrained they were despite the 24 bit
> <br>
> addressing of the 68k.  Is the tcp/ip safe from basic Morris worm shit
> <br>
> (happened while I had that job)?  I&#39;d like to do something internet f=
> acing,
> <br>
> but I have concerns...  Three protocols I&#39;d like to implement for nos=
>  just
> <br>
> for shits and giggles too, so I will be digging in...
> <br>
>
> <br>
> As for the single char per packet thing, there was a discussion about tha=
> t,
> <br>
> iirc, in one of the meeting notes in the IENs, about the time of the cut
> <br>
> from the IMPs.  I forget which though, but it was after WarGames, which i=
> s
> <br>
> also mentioned as a concern in the same or adjacent meeting notes IENs.
> <br>
> I&#39;ll have to look that up.  Maybe they were thinking something simila=
> r, but
> <br>
> to do it otherwise involves compromises, and timeouts, and timeouts could=
>
> <br>
> effect other stuff, waiting n ticks for a packet to maybe fill before
> <br>
> sending any of it...response time issues using both methods, but for
> <br>
> conversational modes one char at a time for both in, echo, and out seems
> <br>
> natural.  Filling a packet first is best for block modes, I would think.
> <br>
>
> <br>
> One of these days I&#39;ll dig into putty.  It needs some terminal emulat=
> ions
> <br>
> though, and I have hundreds of terminal manuals.
> <br>
>
> <br>
> Hell, let&#39;s do a vote.  Most hideously poorly laid out terminal keybo=
> ard.
> <br>
> I vote Hewlett Packard.  Tandem is a distant second.
> <br>
>
> <br>
> On Sun, Jul 2, 2023, 12:20 Kevin Jordan ***@***.***&gt; wrote:
> <br>
>
> <br>
> &gt; I&#39;ve been able to reproduce the problem. Putty is running DtCybe=
> r out of
> <br>
> &gt; input buffers, and buffer exhaustion is handled as a fatal error. Th=
> e
> <br>
> &gt; number of buffers pre-allocated by DtCyber is 1000, so this probably=
>
> <br>
> &gt; indicates that putty is sending one character per Telnet packet. Pos=
> sibly,
> <br>
> &gt; DtCyber can be enhanced to coalesce short packets to avoid the probl=
> em.
> <br>
> &gt; However, as previously suggested, using a file transfer application =
> such as
> <br>
> &gt; FTP is probably a better and more reliable way to upload bulk text t=
> han
> <br>
> &gt; trying to paste into IAF text mode. Even if DtCyber was enhanced to =
> avoid
> <br>
> &gt; the buffer exhaustion problem, IAF TEXT mode would probably abort du=
> e to
> <br>
> &gt; typeahead overrun.
> <br>
> &gt;
> <br>
> &gt; =E2=80=94
> <br>
> &gt; Reply to this email directly, view it on GitHub
> <br>
> &gt; &lt;https://github.com/kej715/DtCyber/discussions/58#discussioncomme=
> nt-6337991&gt;,
> <br>
> &gt; or unsubscribe
> <br>
> &gt; &lt;https://github.com/notifications/unsubscribe-auth/BAPCTQ776WMUJD=
> 3ZUM7BEB3XOG3WNANCNFSM6AAAAAAZ26VCHM&gt;
> <br>
> &gt; .
> <br>
> &gt; You are receiving this because you authored the thread.Message ID:
> <br>
> &gt; ***@***.***&gt;
> <br>
> &gt;
> <br>
>
>
> <p style=3D"font-size:small;-webkit-text-size-adjust:none;color:#666;">&m=
> dash;<br />Reply to this email directly, <a href=3D"https://github.com/ke=
> j715/DtCyber/discussions/58#discussioncomment-6457211">view it on GitHub<=
> /a>, or <a href=3D"https://github.com/notifications/unsubscribe-auth/AATJ=
> 5AF24RWOXNXNTJGFA5DXQLK7RANCNFSM6AAAAAAZ26VCHM">unsubscribe</a>.<br />You=
>  are receiving this because you are subscribed to this thread.<img src=3D=
> "https://github.com/notifications/beacon/AATJ5ADFSDMNR32IWO63AWTXQLK7RA5C=
> NFSM6AAAAAAZ26VCHOWGG33NNVSW45C7OR4XAZNRIRUXGY3VONZWS33OINXW23LFNZ2KUY3PN=
> VWWK3TUL5UWJTQAMKDXW.gif" height=3D"1" width=3D"1" alt=3D"" /><span style=
> =3D"color: transparent; font-size: 0; display: none; visibility: hidden; =
> overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-heig=
> ht: 0; mso-hide: all">Message ID: <span>&lt;kej715/DtCyber/repo-discussio=
> ns/58/comments/6457211</span><span>@</span><span>github</span><span>.</sp=
> an><span>com&gt;</span></span></p>
> <script type=3D"application/ld+json">[
> {
> ***@***.***": "http://schema.org",
> ***@***.***": "EmailMessage",
> "potentialAction": {
> ***@***.***": "ViewAction",
> "target": "https://github.com/kej715/DtCyber/discussions/58#discussioncom=
> ment-6457211",
> "url": "https://github.com/kej715/DtCyber/discussions/58#discussioncommen=
> t-6457211",
> "name": "View Discussion"
> },
> "description": "View this Discussion on GitHub",
> "publisher": {
> ***@***.***": "Organization",
> "name": "GitHub",
> "url": "https://github.com"
> }
> }
> ]</script>=
>
> ----==_mimepart_64b2de78a6ae9_1364c7d0704227ec--

===============================================================================
Ron Young				***@***.***
